pkg/utils/jsonutils_test.go

```diff
@@ -0,0 +1,111 @@
+//
+// Copyright (c) 2015 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package utils
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+	"testing"
+
+	"github.com/heketi/tests"
+)
+
+type testDest struct {
+	Foo string `json:"foo"`
+	Bar int    `json:"bar"`
+}
+
+type dummyReader struct {
+	emsg string
+}
+
+type dummyCloser struct {
+	io.Reader
+}
+
+func (d *dummyReader) Read(p []byte) (n int, err error) {
+	return 0, errors.New(d.emsg)
+}
+
+func (c dummyCloser) Close() error { return nil }
+
+func TestGetJsonFromRequestOk(t *testing.T) {
+	dest := &testDest{}
+	req, err := http.NewRequest(
+		http.MethodPut,
+		"/foo/bar",
+		bytes.NewBufferString(`{"foo": "MyTest", "bar": 88}`))
+	tests.Assert(t, err == nil, "err is nil", err)
+	err = GetJsonFromRequest(req, dest)
+	tests.Assert(t, err == nil, "err is nil", err)
+	tests.Assert(t, dest.Foo == "MyTest",
+		"failed check:", dest.Foo, "==", "MyTest")
+	tests.Assert(t, dest.Bar == 88,
+		"failed check", dest.Bar, "==", 88)
+}
+
+func TestGetJsonFromRequestBadJson(t *testing.T) {
+	dest := &testDest{}
+	req, err := http.NewRequest(
+		http.MethodPut,
+		"/foo/bar",
+		bytes.NewBufferString(`{"foo": `))
+	tests.Assert(t, err == nil, "err is nil", err)
+	err = GetJsonFromRequest(req, dest)
+	tests.Assert(t, err != nil, "err is set", err)
+}
+
+func TestGetJsonFromRequestEmptyBuf(t *testing.T) {
+	dest := &testDest{}
+	req, err := http.NewRequest(
+		http.MethodPut,
+		"/foo/bar",
+		new(bytes.Buffer))
+	tests.Assert(t, err == nil, "err is nil", err)
+	err = GetJsonFromRequest(req, dest)
+	tests.Assert(t, err != nil, "err is set", err)
+}
+
+func TestGetJsonFromRequestBadIo(t *testing.T) {
+	dest := &testDest{}
+	req, err := http.NewRequest(
+		http.MethodPut,
+		"/foo/bar",
+		&dummyReader{"ouch"})
+	tests.Assert(t, err == nil, "err is nil", err)
+	err = GetJsonFromRequest(req, dest)
+	tests.Assert(t, err != nil, "err is set", err)
+}
+
+func TestGetJsonFromResponseOk(t *testing.T) {
+	resp := &http.Response{
+		Status:     "200 OK",
+		StatusCode: 200,
+		Body:       dummyCloser{bytes.NewBufferString(`{"foo": "port", "bar": 22}`)},
+	}
+	dest := &testDest{}
+	err := GetJsonFromResponse(resp, dest)
+	tests.Assert(t, err == nil, "err is nil", err)
+	tests.Assert(t, dest.Foo == "port", "failed check:", dest.Foo, "==", "port")
+	tests.Assert(t, dest.Bar == 22, "failed check", dest.Bar, "==", 22)
```

Comment:
Isn't this kind of message somwhat awkward to read? How about
```
tests.Assert(t, dest.Bar == 22, "Expected Bar == 22, got ", dest.Bar)
```

Reply:
Yes. These are better.

I have some ideas about reducing the boilerplate in the Assert calls, but I wanted to prototype them first, or look around to see if there are other Go libraries that do something similar to what I was thinking. Regardless of my grand plans, I will incorporate these suggestions.